pennylane/qnode.py

```diff
@@ -410,6 +373,163 @@ def construct(self, args, kwargs=None):
+        """Create metric tensor subcircuits for each parameter.
+
+        If the parameter appears in a gate :math:`G`, the subcircuit contains
+        all gates which precede :math:`G`, and :math:`G` is replaced by the variance
+        value of its generator.
+
+        Args:
+            args (tuple): Represent the free parameters passed to the circuit.
+                Here we are not concerned with their values, but with their structure.
+                Each free param is replaced with a :class:`~.variable.Variable` instance.
+
+        Keyword Args:
+            diag_approx (bool): If ``True``, forces the diagonal
+                approximation. Default is ``False``.
+
+        .. note::
+
+            Additional keyword arguments may be passed to the quantum circuit function, however PennyLane
+            does not support differentiating with respect to keyword arguments. Instead,
+            keyword arguments are useful for providing data or 'placeholders' to the quantum circuit function.
+        """
+        # pylint: disable=too-many-statements
+        diag_approx = kwargs.pop("diag_approx", False)
+
+        if not self.ops or not self.cache:
+            # construct the circuit
+            self.construct(args, kwargs)
+
+        # convert the queue to a DAG
+        G = to_DiGraph(self.queue, self.ev)
+
+        # keep track of the layer number
+        layer = 0
+        layer_ops = {0: {"ops": [], "pidx": []}}
+
+        variable_ops_sorted = sorted(list(self.variable_ops.items()), key=lambda x: x[1][0][0])
+
+        for param_idx, gate_param_tuple in variable_ops_sorted:
+            # iterate over all parameters
+            for op_idx, _ in gate_param_tuple:
+                # get all dependents of the existing parameter
+                sub = set(nx.dag.topological_sort(G.subgraph(nx.dag.ancestors(G, op_idx)).copy()))
+
+                # check if any of the dependents are in the
+                # existing layer
+                if set(layer_ops[layer]["ops"]) & sub:
+                    # operation depends on previous layer,
+                    # start a new layer count
+                    layer += 1
+
+                # store the parameters and ops indices for the layer
+                layer_ops.setdefault(layer, {"ops": [], "pidx": []})
+                layer_ops[layer]["ops"].append(op_idx)
+                layer_ops[layer]["pidx"].append(param_idx)
+
+        # iterate through each layer
+        for _, v in layer_ops.items():
+            # for the layer, get the ops and parameter indices
+            ops = v["ops"]
+            param_idx = v["pidx"]
+
+            # get the ops in this layer
+            curr_ops = [op for n, op in G.nodes(data="op") if n in ops]
+
+            # get all ancestor operations
+            ancestors = set()
+            for o in ops:
+                subG = G.subgraph(nx.dag.ancestors(G, o))
+                ancestors |= set(subG.nodes(data="op"))
```

Comment:
This will contain all operations that are direct or indirect ancestors of the current layer, ignoring operations that temporally could be placed after the layer, and going all the way back to the beginning of the circuit